### PR TITLE
Hrcpp 381/cont queries

### PR DESCRIPTION
**Pull request by rigazilla**
This is a continuous queries implementation for the C++ client.
https://issues.jboss.org/browse/HRCPP-381

### CODE REVIEW

File: include/infinispan/hotrod/CacheClientListener.h [hunk 1]
```diff
@@ -37,18 +37,23 @@ class CacheClientListener : public ClientListener
 
     CacheClientListener(RemoteCache<K,V>& cache) : cache((RemoteCacheBase&)cache) {};
     void add_listener(std::function<void(ClientCacheEntryCreatedEvent<K>)> callback) {
+        interestFlag|=0x01;
```
Comment (tristantarrant) [inline include/infinispan/hotrod/CacheClientListener.h]: Maybe move all constants to the top using `const` ?

File: include/infinispan/hotrod/ContinuousQueryListener.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	const std::string& getQuery() const {
+		return query;
+	}
+	std::function<void(ClientCacheEntryCustomEvent)> listenerCustomEvent;
+	CacheClientListener<K, V> cl;
+
+private:
+	std::string query;
+	std::function<void(K, std::tuple<Params...>)> joiningListener;
+    std::function<void(K, std::tuple<Params...>)> leavingListener;
```
Comment (tristantarrant) [inline include/infinispan/hotrod/ContinuousQueryListener.h]: spacing is irregular

File: include/infinispan/hotrod/QueryUtils.h [hunk 3]
```diff
@@ -72,6 +77,30 @@ template <typename T> T unwrapSingleResult(const QueryResponse &qr)
 
 
 #if !defined (_MSC_VER) || (_MSC_VER>=1800)
+
+// typesafe method to turn one projection row of a array of WrappedMessage objs into a tuple
+template <typename H, typename... Params> std::tuple<H, Params...> popTuple(const RepeatedPtrField<WrappedMessage >& wMsgs,  int &k)
+{
+	H s = unwrapSingleValue<H>(wMsgs.Get(k++));
+	std::tuple<Params...> p=popTuple<Params... >(wMsgs,k);
```
Comment (tristantarrant) [inline include/infinispan/hotrod/QueryUtils.h]: spacing around the `=`

File: src/hotrod/impl/protocol/Codec20.cpp [hunk 4]
```diff
@@ -275,6 +275,11 @@ void Codec20::writeClientListenerParams(transport::Transport& t, const ClientLis
     this->writeNamedFactory(t, clientListener.converterFactoryName, converterFactoryParams);
 }
 
+void Codec20::writeClientListenerInterests(transport::Transport& t, unsigned char) const
+{
+	// no op until 2.6 protocol
```
Comment (tristantarrant) [inline src/hotrod/impl/protocol/Codec20.cpp]: I guess no one will invoke it at this version, but maybe throw an exception ?

File: test/ContinuousQueryTest.cpp [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+			{
+		std::cerr << "fail: error in registering .proto model" << std::endl;
+		result = -1;
+		return result;
+	}
+    rMain.add([&metadataCache] {
+        metadataCache.remove("sample_bank_account/bank.proto");
+    });
+    auto *testkm = new BasicTypesProtoStreamMarshaller<int>();
+	auto *testvm = new ProtoStreamMarshaller<sample_bank_account::User>();
```
Comment (tristantarrant) [inline test/ContinuousQueryTest.cpp]: spacing

File: test/SimpleSaslTls.cpp [hunk 6]
```diff
@@ -108,7 +108,7 @@ int main(int argc, char** argv) {
       RemoteCache<std::string, std::string> cache = cacheManager.getCache<std::string, std::string>(km,
           &Marshaller<std::string>::destroy, vm, &Marshaller<std::string>::destroy, std::string("authCache"));
       cacheManager.start();
-      cache.clear();
+        cache.clear();
```
Comment (tristantarrant) [inline test/SimpleSaslTls.cpp]: spurious space ?

File: test/data/standalone.xml [hunk 7]
```diff
@@ -173,6 +173,7 @@
                     <transaction mode="NONE"/>
                 </local-cache>
                 <local-cache name="namedCache" start="EAGER"/>
+                <local-cache name="queryCache" start="EAGER"/>
```
Comment (tristantarrant) [inline test/data/standalone.xml]: we no longer need the `start` attribute

**Comment by tristantarrant** [comment]
Some minor issues. Please squash commits together when you're done

File: src/hotrod/impl/protocol/Codec20.cpp (hunk 4 above)
Comment (rigazilla) [reply src/hotrod/impl/protocol/Codec20.cpp]: No method is used by the add listener operation and needs to be transparent for protocol version lower than 2.6

**Comment by rigazilla** [comment]
@tristantarrant done all the changes but not the one for Codec20, explained above

**Comment by tristantarrant** [comment]
Squashed and pushed

**Pull request closed without merging**